social_core/backends/azuread_oid

```diff
@@ -0,0 +1,137 @@
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+"""
+
+"""
+Azure AD OAuth2 backend, docs at:
+    https://python-social-auth.readthedocs.io/en/latest/backends/azuread.html
+
+See https://nicksnettravels.builttoroam.com/post/2017/01/24/Verifying-Azure-Active-Directory-JWT-Tokens.aspx
+for verifying JWT tokens.
+"""
+
+
+class AzureADOIDOAuth2(AzureADOAuth2):
+    name = "azuread-oid-oauth2"
+    OPENID_CONFIGURATION_URL = "{base_url}/.well-known/openid-configuration{appid}"
+    JWKS_URL = "{base_url}/discovery/keys{appid}"
+
+    @property
+    def tenant_id(self):
+        return self.setting("TENANT_ID", "common")
+
+    def openid_configuration_url(self):
+        return self.OPENID_CONFIGURATION_URL.format(
+            base_url=self.base_url, appid=self._appid()
+        )
+
+    def jwks_url(self):
+        return self.JWKS_URL.format(base_url=self.base_url, appid=self._appid())
+
+    def _appid(self) -> str:
+        return (
+            f"?appid={self.setting('KEY')}" if self.setting("KEY") is not None else ""
+        )
+
+    def get_certificate(self, kid):
+        # retrieve keys from jwks_url
+        resp = self.request(self.jwks_url(), method="GET")
+        resp.raise_for_status()
+
+        # find the proper key for the kid
+        for key in resp.json()["keys"]:
+            if key["kid"] == kid:
+                x5c = key["x5c"][0]
+                break
+        else:
+            raise DecodeError(f"Cannot find kid={kid}")
+
+        return load_der_x509_certificate(base64.b64decode(x5c), default_backend())
+
+    def get_user_id(self, details, response):
+        """Use account oid as unique id."""
+        return response.get("oid")
+
+    def user_data(self, access_token, *args, **kwargs):
+        response = kwargs.get("response")
+        if response and response.get("id_token"):
+            id_token = response.get("id_token")
+        else:
+            id_token = access_token
+
+        # get key id and algorithm
+        key_id = get_unverified_header(id_token)["kid"]
+
+        try:
+            # retrieve certificate for key_id
+            certificate = self.get_certificate(key_id)
+
+            return jwt_decode(
+                id_token,
+                key=certificate.public_key(),  # type: ignore[reportArgumentType]
+                algorithms=["RS256"],
+                audience=self.setting("KEY"),
+            )
+        except (DecodeError, ExpiredSignatureError) as error:
+            raise AuthTokenError(self, error)
```

Comment:
Why is this duplicating so many logic? I would expect that it would only differ in `get_user_id`.